File: leetcode/medium/Powxn.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        """
        Calculates x raised to the power n (x^n).
        This implementation uses the "exponentiation by squaring" (binary exponentiation) algorithm,
        which is efficient with a time complexity of O(log n).

        Args:
            x: The base, a floating-point number.
            n: The exponent, an integer.

        Returns:
            The result of x^n as a floating-point number.
        """
        
        # Initialize the result to 1.
        res = 1.0
        
        # Handle negative exponents: x^(-n) = 1 / (x^n).
        # We convert n to its positive equivalent and remember to invert the result later.
        if n < 0:
            x = 1 / x  # If n is negative, we can invert x and make n positive.
            n = -n     # Convert n to its positive counterpart.
        
        # Perform exponentiation by squaring.
        # This algorithm processes the exponent 'n' bit by bit.
        # If the current bit of 'n' is 1, we multiply 'res' by 'x'.
        # In each step, 'x' is squared, effectively handling higher powers.
        while n > 0:
            # If the least significant bit of 'n' is 1 (n is odd),
            # it means this power of x contributes to the final result.
            if (n & 1) == 1:
                res *= x
            
            # Square x for the next iteration. This prepares x for the next bit position (x^2, x^4, x^8, ...).
            x *= x
            
            # Right-shift n by 1 (equivalent to n = n // 2).
            # This moves to the next bit of the exponent.
            n >>= 1
            
        return res

File: leetcode/medium/Powxn.java
/**
 * Implements a method to compute x raised to the power n ($x^n$).
 *
 * This uses the efficient "Exponentiation by Squaring" (Binary Exponentiation)
 * algorithm, which has a time complexity of O(log |n|) and a space complexity
 * of O(1).
 */
class Solution {
    /**
     * Computes $x^n$ (x raised to the power n).
     *
     * @param x The base number, a double.
     * @param n The exponent, an integer.
     * @return The result of $x^n$.
     */
    public double myPow(double x, int n) {
        // Initialize the result to 1.0.
        double result = 1.0;
        
        // Use a long for the exponent to safely handle the edge case 
        // n = Integer.MIN_VALUE, where -n would overflow a standard int.
        long pow = n;
        
        // 1. Handle Negative Exponents
        // If the exponent is negative, we use the property $x^{-n} = (1/x)^n$.
        if (pow < 0) {
            x = 1.0 / x;
            // Convert to a positive exponent for the main loop.
            pow = -pow; 
        }
        
        // 2. Main Binary Exponentiation Loop
        // The loop terminates when the power becomes 0.
        while (pow > 0) {
            // Check if the least significant bit of 'pow' is 1 (i.e., if 'pow' is odd).
            // If it is odd, it means the current 'x' (which represents $x^{2^k}$ at this step)
            // must be included in the final product.
            if ((pow & 1) == 1) {
                result *= x;
            }
            
            // Square the base: $x \to x^2 \to x^4 \to x^8 \dots$
            // This prepares 'x' for the next bit of the exponent.
            x *= x;
            
            // Right shift 'pow' by 1: $n \to n/2 \to n/4 \dots$
            // This effectively moves to the next higher-order bit of the exponent.
            pow >>= 1;
        }
        
        return result;
    }
}